lib/opencensus/trace/exporters/ocagent.rb

```diff
@@ -15,14 +15,270 @@
+        # @return [Boolean] If steam is not started or stop returns true.
+        #  If trace export stream alreay running return false.
+        def start
+          return false unless @stopped
+
+          @trace_req_queue = TraceEnumeratorQueue.new @trace_stream_sleep_delay
+          Thread.new { background_export_run }
+          @stopped = false
+          true
+        end
+
+        # Export spans to OpenCensus Agent service.
+        #
+        # @param [Array<OpenCensus::Trace::Span>] spans The captured spans to
+        #   export to OpenCensus Agent service
+        #
+        def export spans
+          return nil if spans.nil? || spans.empty?
+
+          start unless @trace_req_queue
+
+          return if stopped?
+
+          converter = Converter.new @resource
+          req = AgentProto::Trace::V1::ExportTraceServiceRequest.new(
+            node: node_info,
+            resource: resource,
+            spans: spans.map { |span| converter.convert_span span }
+          )
+          @trace_req_queue.push req
+
+          nil
+        end
+
+        private
+
+        # rubocop:disable Metrics/MethodLength
+
+        # Create node info proto.
+        # @return [OpenCensus::Proto::Agent::Common::V1::Node]
+        #
+        def create_node_info
+          time = Time.now.utc
+          timestamp = Google::Protobuf::Timestamp.new(
+            seconds: time.to_i,
+            nanos: time.nsec
+          )
+
+          identifier = AgentProto::Common::V1::ProcessIdentifier.new(
+            host_name: Socket.gethostname,
+            pid: Process.pid,
+            start_timestamp: timestamp
+          )
+
+          library_info = AgentProto::Common::V1::LibraryInfo.new(
+            language: AgentProto::Common::V1::LibraryInfo::Language::RUBY,
+            exporter_version: OpenCensus::OCAgent::VERSION,
+            core_library_version: OpenCensus::VERSION
+          )
+
+          service_info = AgentProto::Common::V1::ServiceInfo.new(
+            name: @service_name
+          )
+
+          OpenCensus::Proto::Agent::Common::V1::Node.new(
+            identifier: identifier,
+            library_info: library_info,
+            service_info: service_info
+          )
+        end
+
+        # rubocop:enable Metrics/MethodLength
+
+        # Create resouce proto object.
+        #
+        # @param [String] type Resource type.
+        #   Resource that is associated with each span which is going to export
+        # @param [Hash<String, String>] labels Set of labels that describe the
+        #   resource. Optional.
+        # @return [OpenCensus::Proto::Resource::V1::Resource]
+        #
+        def create_resource type, labels: nil
+          options = { type: type }
+          options[:labels] = labels if labels
+          OpenCensus::Proto::Resource::V1::Resource.new options
+        end
+
+        # Run exporter stream.
+        def background_export_run
+          client.export(@trace_req_queue.each_item).each {}
```

Comment:
Why do you need to call `.each{}` on the result?

Reply:
export is bidirectional grpc call which need to be kept alive for the entire life of process and export returning Enumerator instance with empty objects.